src/System.Management.Automation/engine/parser/tokenizer.cs

```diff
@@ -3289,24 +3331,58 @@ private static bool TryGetNumberValue(string strNum, bool hex, bool real, char s
                         {
                             multiplier = -multiplier;
                         }
+
                         strNum = strNum.Substring(1);
                     }
+
                     style = hex ? NumberStyles.AllowHexSpecifier : NumberStyles.AllowLeadingSign;
 
                     long longValue;
-                    if (suffix == 'l' || suffix == 'L')
+                    switch (suffix)
                     {
-                        if (long.TryParse(strNum, style, NumberFormatInfo.InvariantInfo, out longValue))
-                        {
-                            result = longValue * multiplier;
-                            return true;
-                        }
-                        result = null;
-                        return false;
+                        case NumberSuffixFlags.Long:
+                            if (long.TryParse(strNum, style, NumberFormatInfo.InvariantInfo, out longValue))
+                            {
+                                result = longValue * multiplier;
+                                return true;
+                            }
+
+                            result = null;
+                            return false;
+                        case NumberSuffixFlags.Short:
+                            if (short.TryParse(strNum, style, NumberFormatInfo.InvariantInfo, out short s))
+                            {
+                                result = (short)(s * multiplier);
+                                return true;
+                            }
+
+                            result = null;
+                            return false;
+                        default:
+                            if (suffix.HasFlag(NumberSuffixFlags.Unsigned) && ulong.TryParse(strNum, style, NumberFormatInfo.InvariantInfo, out ulong u))
+                            {
+                                u *= (ulong)multiplier;
+                                if (suffix.HasFlag(NumberSuffixFlags.Short) && u <= ushort.MaxValue)
+                                {
+                                    result = (ushort)u;
+                                }
+                                else if (!suffix.HasFlag(NumberSuffixFlags.Long) && u <= uint.MaxValue)
+                                {
+                                    result = (uint)u;
+                                }
+                                else
+                                {
+                                    // ulong
+                                    result = u;
+                                }
+
+                                return true;
```

Comment:
Comparing this switch statement and the one above, this one is not as thorough as the above one. The above one takes into account the cases when `suffix` is an invalid combination, such as `None | Unsigned`, but this one doesn't.

I'm not saying that we should handle those __impossible__ values for `suffix`, but more about consistency. I'm fine that we assume the value of `suffix` is always legit, and simply the above switch statement to:
```
        switch (suffix)
        {
            case NumberSuffixFlags.Long:
                ...
            case NumberSuffixFlags.Short:
                ...
            case NumberSuffixFlags.Unsigned | NumberSuffixFlags.Long:
                ...
            case NumberSuffixFlags.Unsigned | NumberSuffixFlags.Short:
                ...
            case NumberSuffixFlags.Unsigned:
                ...
            default:
                // NumberSuffixFlags.None
                result = d * multiplier;
                break;
        }
```

Reply:
Hmm. I see what you mean. I'll have it error out on invalid combinations; even if the user can't do something like `None | Unsigned` normally I'd rather error on the side of caution and refuse to parse there.

Reply:
Restructured this similarly to the first switch.

`None` flags case now breaks the switch, while `default` fails the parse. Split out the `uint` into several separate cases. 😄